File: java/src/test/java/com/alxarafe/domain/inventory/valueobject/QuantityTest.java
package com.alxarafe.domain.inventory.valueobject;

import com.alxarafe.app.domain.inventory.valueobject.Quantity;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class QuantityTest {

    @Test
    void create() {
        var qty = Quantity.fromDecimal(10.0, "EA");
        assertEquals(10.0, qty.value());
        assertEquals("EA", qty.unit());
    }

    @Test
    void zeroValue() {
        assertThrows(IllegalArgumentException.class, () -> Quantity.fromDecimal(0.0, "EA"));
    }

    @Test
    void negativeValue() {
        assertThrows(IllegalArgumentException.class, () -> Quantity.fromDecimal(-5.0, "EA"));
    }

    @Test
    void nullUnit() {
        assertThrows(NullPointerException.class, () -> Quantity.fromDecimal(1.0, null));
    }

    @Test
    void blankUnit() {
        assertThrows(IllegalArgumentException.class, () -> Quantity.fromDecimal(1.0, ""));
    }

    @Test
    void add() {
        var result = Quantity.fromDecimal(5.0, "EA").add(Quantity.fromDecimal(3.0, "EA"));
        assertEquals(8.0, result.value());
        assertEquals("EA", result.unit());
    }

    @Test
    void addDifferentUnit() {
        var a = Quantity.fromDecimal(5.0, "EA");
        assertThrows(IllegalArgumentException.class, () -> a.add(Quantity.fromDecimal(3.0, "KG")));
    }

    @Test
    void subtract() {
        var result = Quantity.fromDecimal(5.0, "EA").subtract(Quantity.fromDecimal(3.0, "EA"));
        assertEquals(2.0, result.value());
    }

    @Test
    void subtractInsufficient() {
        var a = Quantity.fromDecimal(3.0, "EA");
        assertThrows(IllegalArgumentException.class, () -> a.subtract(Quantity.fromDecimal(5.0, "EA")));
    }

    @Test
    void subtractDifferentUnit() {
        var a = Quantity.fromDecimal(5.0, "EA");
        assertThrows(IllegalArgumentException.class, () -> a.subtract(Quantity.fromDecimal(3.0, "KG")));
    }

    @Test
    void equals() {
        assertEquals(Quantity.fromDecimal(10.0, "EA"), Quantity.fromDecimal(10.0, "EA"));
        assertNotEquals(Quantity.fromDecimal(10.0, "EA"), Quantity.fromDecimal(5.0, "EA"));
    }

    @Test
    void scaledRoundTrip() {
        var qty = Quantity.fromScaled(30_500_001L, "EA");
        assertEquals(30_500_001L, qty.scaledUnits());
        assertEquals(30.500001, qty.value());
    }

    @Test
    void decimalStringRoundTrip() {
        var qty = Quantity.fromDecimalString("30.50025", "EA");
        assertEquals(30_500_250L, qty.scaledUnits());
        assertEquals("30.500250", qty.toDecimalString());
        assertEquals(30.50025, qty.value());
    }

    @Test
    void exactAdditionUsesScaledUnits() {
        var sum = Quantity.fromDecimal(0.1, "EA").add(Quantity.fromDecimal(0.2, "EA"));
        assertEquals(0.3, sum.value());
    }

    @Test
    void roundsHalfUpOnBoundary() {
        var qty = Quantity.fromDecimal(2.5000004, "EA");
        assertEquals(2_500_000L, qty.scaledUnits());
        assertEquals(2.5, qty.value());
    }

    @Test
    void rejectsNonFinite() {
        assertThrows(IllegalArgumentException.class, () -> Quantity.fromDecimal(Double.NaN, "EA"));
        assertThrows(IllegalArgumentException.class, () -> Quantity.fromDecimal(Double.POSITIVE_INFINITY, "EA"));
    }

    @Test
    void rejectsInvalidDecimalString() {
        assertThrows(IllegalArgumentException.class, () -> Quantity.fromDecimalString("abc", "EA"));
    }
}

File: php/tests/Unit/Domain/Inventory/ValueObject/QuantityTest.php
<?php

declare(strict_types=1);

namespace Tests\Unit\Domain\Inventory\ValueObject;

use Alxarafe\App\Domain\Inventory\ValueObject\Quantity;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;

final class QuantityTest extends TestCase
{
    public function testCreate(): void
    {
        $qty = Quantity::fromDecimal(10.0, 'EA');
        self::assertSame(10.0, $qty->value());
        self::assertSame('EA', $qty->unit());
    }

    public function testZeroValue(): void
    {
        $this->expectException(InvalidArgumentException::class);
        Quantity::fromDecimal(0.0, 'EA');
    }

    public function testNegativeValue(): void
    {
        $this->expectException(InvalidArgumentException::class);
        Quantity::fromDecimal(-5.0, 'EA');
    }

    public function testEmptyUnit(): void
    {
        $this->expectException(InvalidArgumentException::class);
        Quantity::fromDecimal(1.0, '');
    }

    public function testAdd(): void
    {
        $a = Quantity::fromDecimal(5.0, 'EA');
        $b = Quantity::fromDecimal(3.0, 'EA');
        $result = $a->add($b);
        self::assertSame(8.0, $result->value());
        self::assertSame('EA', $result->unit());
    }

    public function testAddDifferentUnit(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $a = Quantity::fromDecimal(5.0, 'EA');
        $b = Quantity::fromDecimal(3.0, 'KG');
        $a->add($b);
    }

    public function testSubtract(): void
    {
        $a = Quantity::fromDecimal(5.0, 'EA');
        $b = Quantity::fromDecimal(3.0, 'EA');
        $result = $a->subtract($b);
        self::assertSame(2.0, $result->value());
    }

    public function testSubtractInsufficient(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $a = Quantity::fromDecimal(3.0, 'EA');
        $b = Quantity::fromDecimal(5.0, 'EA');
        $a->subtract($b);
    }

    public function testSubtractDifferentUnit(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $a = Quantity::fromDecimal(5.0, 'EA');
        $b = Quantity::fromDecimal(3.0, 'KG');
        $a->subtract($b);
    }

    public function testEquals(): void
    {
        $a = Quantity::fromDecimal(10.0, 'EA');
        $b = Quantity::fromDecimal(10.0, 'EA');
        $c = Quantity::fromDecimal(5.0, 'EA');
        self::assertTrue($a->equals($b));
        self::assertFalse($a->equals($c));
    }

    public function testScaledRoundTrip(): void
    {
        $qty = Quantity::fromScaled(30_500_001, 'EA');
        self::assertSame(30_500_001, $qty->scaledUnits());
        self::assertSame(30.500001, $qty->value());
    }

    public function testDecimalStringRoundTrip(): void
    {
        $qty = Quantity::fromDecimalString('30.50025', 'EA');
        self::assertSame(30_500_250, $qty->scaledUnits());
        self::assertSame('30.500250', $qty->toDecimalString());
        self::assertSame(30.50025, $qty->value());
    }

    public function testExactAdditionUsesScaledUnits(): void
    {
        $a = Quantity::fromDecimal(0.1, 'EA');
        $b = Quantity::fromDecimal(0.2, 'EA');
        // 0.1 + 0.2 no debe arrastrar el error binario del float.
        self::assertSame(0.3, $a->add($b)->value());
    }

    public function testRoundsHalfUpOnBoundary(): void
    {
        $qty = Quantity::fromDecimal(2.5000004, 'EA');
        self::assertSame(2_500_000, $qty->scaledUnits());
        self::assertSame(2.5, $qty->value());
    }

    public function testRejectsNonFinite(): void
    {
        $this->expectException(InvalidArgumentException::class);
        Quantity::fromDecimal(NAN, 'EA');
    }

    public function testRejectsInfinite(): void
    {
        $this->expectException(InvalidArgumentException::class);
        Quantity::fromDecimal(INF, 'EA');
    }

    public function testRejectsInvalidDecimalString(): void
    {
        $this->expectException(InvalidArgumentException::class);
        Quantity::fromDecimalString('abc', 'EA');
    }
}
